grisp/grisp_base/common/build/nifs/grisp_nspi_nif.c

```diff
@@ -0,0 +1,137 @@
+#define STATIC_ERLANG_NIF 1
+
+#include <assert.h>
+#include <bsp.h>
+#include <bsp/atsam-spi.h>
+#include <bsp/spi.h>
+#include <dev/spi/spi.h>
+#include <erl_nif.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <rtems.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#include "sys.h"
+
+#define CPOL_LOW 0
+#define CPOL_HIGH 1
+#define CPHA_LEADING 0
+#define CPHA_TRAILING 2
+
+/* Make sure to keep this at sync with the -define(res_max_size.. in spi.erl */
+#define RES_MAX_SIZE 256
+
+static ERL_NIF_TERM atom_ok;
+static ERL_NIF_TERM atom_error;
+static ErlNifResourceType *spi_data;
+
+typedef struct {
+    int fd;
+} grisp_spi_data;
+
+static const atsam_spi_config spi_config = {.spi_peripheral_id = ID_SPI0,
+                                            .spi_regs = SPI0};
+
+static void spi_data_dtor(ErlNifEnv *env, void *obj) {
+    close(((grisp_spi_data *)obj)->fd);
+}
+
+int load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info) {
+    atom_ok = enif_make_atom(env, "ok");
+    atom_error = enif_make_atom(env, "error");
+    spi_data = enif_open_resource_type(env, NULL, "spi_data",
+                                       (ErlNifResourceDtor *)spi_data_dtor,
+                                       ERL_NIF_RT_CREATE, NULL);
+
+    assert(spi_data != NULL);
+
+    int rv;
+
+    /* bus registration */
+    rv = spi_bus_register_atsam(ATSAM_SPI_0_BUS_PATH, &spi_config);
+
+    if (rv != 0) perror("SPI bus registration failed");
+    assert(rv == 0);
+
+    return 0;
+}
+
+static ERL_NIF_TERM error_tuple(ErlNifEnv *env, char *reason) {
+    return enif_make_tuple2(env, atom_error, enif_make_atom(env, reason));
+}
+
+static ERL_NIF_TERM open_spi(ErlNifEnv *env, int argc,
+                             const ERL_NIF_TERM argv[]) {
+    ERL_NIF_TERM ret;
+    uint32_t speed = 100000;
+
+    printf(stderr, "Tying to alloc resource");
+    grisp_spi_data *data =
+        enif_alloc_resource(spi_data, sizeof(grisp_spi_data));
+
+    data->fd = open(ATSAM_SPI_0_BUS_PATH, O_RDWR);
+
+    if (data->fd == -1) return error_tuple(env, "open_spi");
```

Comment:
I think it's best to always use brackets around if statements, so that the intention of the code is clear (not sure if `clang-format` adds this?).